zip/zlib.nim

```diff
@@ -260,129 +262,111 @@ proc compress*(sourceBuf: string; level=Z_DEFAULT_COMPRESSION; stream=GZIP_STREA
-  if space < left:
-    space = left
-
-  var decompressed = newStringOfCap(space)
-
-  # Initialize output.
-  z.nextOut = addr(decompressed[0])
-  # Output generated so far.
-  var have = 0
-
-
-  var windowBits = MAX_WBITS
-  case (stream)
-  of RAW_DEFLATE: windowBits = -MAX_WBITS
-  of ZLIB_STREAM: windowBits = MAX_WBITS
-  of GZIP_STREAM: windowBits = MAX_WBITS + 16
-  of DETECT_STREAM: windowBits = MAX_WBITS + 32
-
-  z.availIn = 0;
-  var status = inflateInit2(z, windowBits.int32)
+  var strm: ZStream
+  var decompressed: string = ""
+  var sbytes = 0
+  var wbytes = 0
+  ##  allocate inflate state
+
+  strm.availIn = 0
+  var wbits = case (stream)
+  of RAW_DEFLATE:  -MAX_WBITS
+  of ZLIB_STREAM:   MAX_WBITS
+  of GZIP_STREAM:   MAX_WBITS + 16
+  of DETECT_STREAM: MAX_WBITS + 32
+
+  var status = inflateInit2(strm, wbits.int32)
   if status != Z_OK:
-    # Out of memory.
     return
 
-  # Make sure memory allocated by inflateInit2() is freed eventually.
-  defer: discard inflateEnd(z)
-
-  # Decompress all of self.
+  # run loop until all input is consumed.
+  # handle concatenated deflated stream with header.
   while true:
-    # Allow for concatenated gzip streams (per RFC 1952).
-    if status == Z_STREAM_END:
-      discard inflateReset(z)
+    strm.availIn = (sourceLen - sbytes).int32
 
-    # Provide input for inflate.
-    if z.availIn == 0:
-      # This only makes sense in the C version using unsigned values.
-      z.availIn = left
-      left -= z.availIn
+    # no more input available
+    if (sourceLen - sbytes) <= 0: break
+    strm.nextIn = sourceBuf[sbytes].unsafeaddr
 
-    # Decompress the available input.
+    #  run inflate() on available input until output buffer is full
     while true:
-      # Allocate more output space if none left.
-      if space == have:
-        decompressed.setLen(space)
-        # Double space, handle overflow.
-        space = space shl 1
-        if space < have:
-          # Space was likely already maxed out.
-          discard inflateEnd(z)
-          return
-
-        # Increase space.
-        decompressed.setLen(space)
-        # Update output pointer (might have moved).
-        z.nextOut = addr(decompressed[have])
-
-      # Provide output space for inflate.
-      z.availOut = zlib.Uint(space - have)
-      have += z.availOut;
-
-      # Inflate and update the decompressed size.
-      status = inflate(z, Z_SYNC_FLUSH);
-      have -= z.availOut;
-
-      # Bail out if any errors.
-      if status != Z_OK and status != Z_BUF_ERROR and status != Z_STREAM_END:
-        # Invalid gzip stream.
-        discard inflateEnd(z)
+      # if written bytes >= output size : resize output
+      if wbytes >= decompressed.len:
+        let cur_outlen = decompressed.len
+        let new_outlen = if decompressed.len == 0: sourceLen*2 else: decompressed.len*2
+        if new_outlen < cur_outlen: # unsigned integer overflow, buffer too big
+          discard inflateEnd(strm); return
```

Comment:
I think it would be a good idea, in fact I had already thought about it, but it would change the original API. Unless you mean raising an error, which would be a good idea but would brusquely change the original API.
I think the python stdlib also does this.
It would likely work since this call is unlikely to be a bottleneck for an algorithm.
Anyone for example uncompressing/compressing from a file would likely implement their own loop for best performance.